ui/app/src/views/SwapPage/components/SlippageTolerance.tsx

```diff
@@ -4,6 +4,29 @@ export const SlippageTolerance = (props: {
   slippage: string;
   onUpdate: (val: string) => any;
 }) => {
+  const handleOnInput = (e: Event) => {
+    let v = (e.target as HTMLInputElement).value;
+
+    if (isNaN(parseFloat(v)) || parseFloat(v) < 0) {
+      v = "0";
+    }
+
+    const shouldTrimInput = parseFloat(v) > 999999999;
+    const inputContainsManitssa = v.includes(".");
+
+    // if value is an integer, slice off last digit
+    if (shouldTrimInput && !inputContainsManitssa) {
+      v = v.slice(0, -1);
+    } else if (shouldTrimInput && inputContainsManitssa) {
+      // trim value to an integer of 9 digits
+      const decimalIndex = v.indexOf(".");
+      v = v.slice(0, decimalIndex);
+    }
+
+    // update the input field value
+    (e.target as HTMLInputElement).value = v;
```

Comment:
agree that presentational components should be kept as pure as possible.  

Things get interesting when the e.target.value > max allowable data.slippage.value. The e.target.value will continue to increase (and display to user the additional characters), even though the data.slippage.value stays under the max allowable limit, even though the value of data.slippage is bound to the value of the input field.

Part of the struggle is that the onInput event fires after the input field's value has already updated to whatever was entered.

I took in a few considerations on how to resolve this:
1. use the input type="number"'s max attribute to limit input character length, but determined this path won't be fruitful since we're taking into account different max digits before and after the decimal place.
2. Place a key on the input field that increments whenever e.target.value > props.slippage, which will rerender the entire input field, and set reset the value to the latest props.slippage.  This works, but the input element loses focus whenever the rerender occurs, and therefore the caret position within the input element is lost as well.
3. Add an onBeforeInput event that runs a few checks against the characters the user is trying to input, and if the checks fail, the character(s) won't be added to e.target.value via running e.preventDefault(), and therefore the onInput event won't fire as well.  

I went down this path 3, and the new shouldPreventInput() method was born. It takes in the parameters:
- input field's value before the input event
- the position of the caret in the text box at the time as to determine if the user is manipulating the significand or the mantissa
- the character(s) the user is trying to add to e.target.value
- the max allowed whole number
- the max allowed numbers after decimal

shouldPreventInput uses these arguments to ensure the user won't:
- add too many digits to the whole number
- add too many digits after the decimal
- copy and paste large values on either side of the decimal
- enter in a non-numeric value that contains more than one decimal point
- enter in negative values

The input fields were converted to type="text" because I needed access to e.target.selectionStart, which isn't available on input type="number". e.target.selectionStart gives you the position of the caret in the input field, so I could tell whether the user was trying to add a character before or after the decimal place. and then prevent that action if it was higher than max allowable on either side. all of the checks to only allow numbers are built into the shouldPreventInput method, but please feel free to stress test it and lmk if would like some mods.